### DISCUSSION TITLE
Syntax error when extending glz::json_t with glm types

### DISCUSSION THREAD

**Discussion by exdal** [stephenberry/glaze#1333] [Q&A] [answered]
Hey, I've been trying to customize `glz::json_t` for my needs, I implemented exact copy of `glz::json_t` but with custom `glz::json_t::val_t` and here is how it looks:
```cpp
using val_t = std::variant<null_t, std::string, bool, array_t, object_t, double, glm::vec2>;

// ...
// Meta impl. for my custom types, outside of 'MyCustomJson' struct
template<>
struct glz::meta<glm::vec2> {
    static constexpr std::string_view name = "glm::vec2";
    using T = glm::vec2;
    static constexpr auto value = object(&T::x, &T::y);
};
```

For some reason, [variant_is_auto_deducible](https://github.com/stephenberry/glaze/blob/6d249bf40f926571c54fd16f3c4ebb3b12f8060e/include/glaze/json/read.hpp#L2636) always evaluates to false thus causing a weird error like `ec: syntax_error, location: 0`. I need to use `glz::json_t` because the data I am trying to read is not predictable (it's a scene deserialization with ECS stuff). Are there any ways to make `glm::vec2` differ from, in my guess, `array_t` or `object_t`?

**Comment by exdal** [comment]
Exact location of where syntax error is set at: https://github.com/stephenberry/glaze/blob/6d249bf40f926571c54fd16f3c4ebb3b12f8060e/include/glaze/json/read.hpp#L447

`it` argument is a string at position 0 of my json string.

**Comment by stephenberry** [comment, marked as answer]
The parsing is failing there because the variant isn't auto-deducible and thus defaults to `null_t` and tries to parse into this `null_t`, which is why it is erroring in the specialization for `always_null_t`.

You're correct that your meta for `glm::vec2` is conflicting with `object_t`, since both are objects (`meta<glm::vec2>`: `object(&T::x, &T::y);`).

The easiest solution is probably to just use `glz::json_t` and then you can call `if (value.contains("x") && value.contains("y"))` this should indicate that you have a `glm::vec2`. There is an active issue open to support reading a `glz::json_t` value into a concrete C++ struct [here](https://github.com/stephenberry/glaze/issues/1164). This would be useful for you as well. Currently you would have to write the value to JSON and then read it into a concrete `glm::vec2`.

I also think I should add a `dump` function to `glz::json_t` so that once you find your target value you can easily dump it to a string.

**Reply by exdal** [reply]
Thank you, I will be following that issue.
